.pre-commit-config.yaml

```diff
@@ -1,16 +1,13 @@
 repos:
 - repo: https://github.com/pre-commit/pre-commit-hooks
-  rev: v4.2.0
+  rev: v4.5.0
   hooks:
     - id: check-yaml
     - id: end-of-file-fixer
     - id: trailing-whitespace
-- repo: https://github.com/pycqa/flake8
-  rev: 6.1.0
+- repo: https://github.com/astral-sh/ruff-pre-commit
+  rev: v0.2.1
   hooks:
-    - id: flake8
-- repo: https://github.com/psf/black
-  rev: 23.7.0
-  hooks:
-    - id: black
-      language_version: python
+    - id: ruff
+      args: [--fix, --exit-non-zero-on-fix]
```

Comment:
Sorry why do we need `--exit-non-zero-on-fix`?

Reply:
Otherwise it'll exit status code zero, which means pre-commit will merrily let you land the changes with outstanding linting issues. This way the commit fails, you review & add the auto-fixes that ruff has made, and you try the commit again.

Reply:
> merrily let you land the changes with outstanding linting issues.

But in this case the changes are fixed, no?

Reply:
It turns out that pre-commit fails if a hook changes any of the files that the user has staged (without then doing a git add) so I agree that this looks superfluous. I'll remove it.

Reply:
Hmmm. I say that, but...

We're running this in two ways:
 * In the local developer's git repo, _only if they have pre-commit set up_.
 * In GitHub Actions, which enforces linting regardless of a dev's local setup.

So we do need this to exit non-zero really, for GitHub Actions to fail the build if there are linting fixes. We don't want them to silently get applied within the CI env but never committed back and for it to pretend it's all fine.

If someone commits something without pre-commit set up on their local repo, then when you run this check in CI, it will fix the files and the job will exit status code zero.

Reply:
I think the easiest thing to do here would be to:

1. Keep the precommit with just `args: [--fix]`
2. Instead of running the precommit GHA, run ruff directly via any of https://docs.astral.sh/ruff/integrations/#github-actions